### ISSUE TITLE
Application Layer / Architecture including OTA, etc. for Contiki (interest poll)

### ISSUE THREAD

**Issue by joakimeriksson** [contiki-os/contiki#1035]
I have been working quite a while at Yanzi networks developing Contiki software and now we are considering to upstream some of the application layer software and extensions that have been developed. My question is if there are any interests in a complete application architecture implementation in upstream Contiki?

It includes support for
- Discovery of functionality of devices
- OTA for full firmware image upgrade
- Read/write variables in the application object / instance model.

This application architecture have been ported to STM32W and CC2538 platforms
(and also tested on smaller platforms) and is based on a UDP protocol.

Included in this is also a set of scripts, libraries and tools for doing discovery, OTA, etc. And full documentation of the protocol and application architecture.

**Comment by alignan** [comment]
:+1:

**Comment by jbnhf23** [comment]
indeed. Could be awesome. Reinventing the wheel, is not what I prefer ;-)

**Comment by areaarea** [comment]
That would be awesome!

Adam

**Comment by g-oikonomou** [comment]
:+1:

**Comment by aguirrem** [comment]
:+1:

**Comment by joakimeriksson** [comment]
Ok, so thanks for the thumbs - I will start lobbying a bit to get this into shape for open-sourcing then. We are also working on some other cool and useful things that I will try to get upstreamed!

**Comment by bradjc** [comment]
:+1:

**Comment by sdawans** [comment]
:+1:

I would try to make OTA as modular as possible - the update logic and underlying code dissemination strategies are vast

**Comment by jonnteolsson** [comment]
Utilizing this and an open standards like OMA LWM2M on top of COAP would be really nice.

**Comment by joakimeriksson** [comment]
Totally agree! But the first version (the one we already have) will be using another very similar approach (almost the same, but before OMA LWM2M suggested their object model). And I guess LWM2M is still in some need for getting completely final before implementing it fully.

**Comment by areaarea** [comment]
My druthers would be to get an OTA option in the tree but to make sure it is factored in a way that accommodates logical upgrades that we can see coming now (e.g. OMA LWM2M) and normal separations (network layers and such).  But getting something in place that can be improved would be a huge step forward IMO.

**Comment by dazhbog** [comment]
:+1:

**Comment by troymccann** [comment]
Are there any updates for this? I'm looking into OTA firmware updates with the CC2538 with Contiki at the moment; it would be great to avoid reinventing the wheel!

**Comment by joakimeriksson** [comment]
Yes, this is work in progress - we are preparing a pull request for the application layer soon. With luck it will be a PR later this week.

**Comment by troymccann** [comment]
Excellent, thanks @joakimeriksson ! Very much appreciated. :)

**Comment by MaximilianWorthZeroMillion** [comment]
Hey @joakimeriksson,
your work is highly appreciated! What is your current status?

**Comment by joakimeriksson** [comment]
Sorry for being a bit slow. We are preparing the PR's for this right now - but also a bit depend on the new 6LoWPAN adaptation layer PR is getting in there. But we will try to make these PR's that depend on that PR one if it is not pulled anytime soon.

**Comment by jonnteolsson** [comment]
any updates, we see great interest in such a feature!

**Comment by joakimeriksson** [comment]
Push any of the mergers to pull the 6LoWPAN adaptation layer PR and we can try to upstream this ASAP ;-)

**Comment by JelmerT** [comment]
Looking forward to this one :+1:

**Comment by marco-pistone** [comment]
:+1:

**Comment by knowic** [comment]
Thanks，we are looking forward to it
....

**Comment by joakimeriksson** [comment]
Yes. Now when the adaptation layer is in we will start pushing this soon!

**Comment by arurke** [comment]
:+1:

**Comment by tonnenpinguin** [comment]
Any updates on this @joakimeriksson? ;-)

**Comment by joakimeriksson** [comment]
We are working on it - we might start pointing you to our "work-in-progress" branch so that you could do early tests. Sorry for being slow on this - I have spent a lot of time fixing other issues. But I it on the top-list of our priorities.

**Comment by tonnenpinguin** [comment]
Thanks for the update!
Of course we could contribute some testing/bug reporting!

**Comment by lligios** [comment]
Thank you @joakimeriksson 
We could also contribute for testing and development.

**Comment by msolters** [comment]
Anything testable yet?

**Comment by knowic** [comment]
Is there any clear time budget about this？ We are looking forward it.

**Comment by joakimeriksson** [comment]
No, clear time budget - but we will work on this in a few days - hopefully able to get a testable version for anyone to test before the weekend!

**Comment by msolters** [comment]
bump :)

**Comment by LanshunNie** [comment]
Is there any update? We are looking forward to OTA, : )

**Comment by joakimeriksson** [comment]
Yes, we are working on it - got it to a level where we have one platform supported (a CC2538 based) but are adding Zolertia Re-Mote now so that there is an official Contiki platform to test with also. We will work with it today but I do not expect it to be fully ready today.

**Comment by LanshunNie** [comment]
Awesome!

**Comment by joakimeriksson** [comment]
Ok, anyone who like to test can find the pre-release of the Sparrow application layer and tools here:
https://github.com/sics-iot/sparrow

Includes bootloader, OTA-tools, new NBR / Serial Radio, etc.
